### PR TITLE
Refactoring and API

### PR DESCRIPTION
**Pull request by encukou**
Here's a start for #494.
I recommend reading the full diff instead of the individual commits. Squash when ready.

If you'd like to review this, let me know in the comments, so I don't merge without your input.

## Features/Changes
* A JSON API with two flavors: input (what we get from "forks") and output (available at at `/v0/naucse.json`)
  * The API is described and validated using jsonschema.
  * Uses HTML for content, never Markdown.
  * Relative URLs use a custom `naucse:` schema for the API. This gets converted in the HTML "sanitization" step.
* The part that should be in "forks" is moved to the `naucse_render` module, which should become its own repo & PyPI package.
* When one course's build fails, everything fails. No silent "an error occured" pages.
  * Only fully applies in "freeze mode" (with `NAUCSE_FREEZE=1`, or Flask in non-debug mode).
  * In debug mode, lessons are rendered on demand instead of at startup. 
* Doesn't do fancy guessing of repo URL&branch for edit info. Use `NAUCSE_MAIN_REPO_URL` and `NAUCSE_MAIN_REPO_BRANCH` if needed (e.g. on Travis).
* The model is better defined. Notably:
  * A *course* has a list of *sessions*
  * A session has a list of *materials*
  * A material may be for a *lesson* (or an external link)
  * A course (!) has *lessons* (which are referenced by materials). (All lessons are in a course; there's a special hidden `lessons` course containing the "local"/"canonical" lessons.)
* The distinction between forks and local content is disappearing – I'd like to get rid of local content eventually.
* The distinction between "courses" and "runs" in the model is disappearing: mostly they just have different slugs (e.g. `courses/mi-pyt` vs. `2018/pyladies-brno-jaro`) and are in different collections for lookup.
* Instead of `FORKS_ENABLED=true`, there's `NAUCSE_TRUSTED_REPOS='*'`. You can filter allowed repos a bit (docs TODO).
* The model is different (`converters.py` is utilities for stuff you can load/dump from/to a versioned JSON API).
* Pages request LaTeX using "`modules: {'katex': '0.7.1'}`. This should become a general mechanism to request some extra functionality (like JS/CSS).
* Model objects have info about their URLs.

### Injects code into arca forks
Currently, forks of naucse don't export structured data; almost everything is converted to HTML on the fork side.  This makes getting e.g. a list of sessions quite hard. This PR adds a nasty hack that injects code into the container (in `arcs_renderer.py`).
My plan is like to install `naucse_render=0.1` into the Arca containers instead of injecting the code. Looking for better solutions.

### API v0
The API is at not finalized, but should be useful. It's "version 0", i.e. please coordinate if you use it.

### naucse_render
I intend to use `naucse_render == 0.1` as a one-off for "legacy" forks. Once there's an API, this can be beautified together with the source data.

### Ugly templates

I kept changes in the result minimal, but did refactor templates a bit. Whitespace/indentation looks weird sometimes. 

## TODO
* [ ] Unit tests
  * [ ] Fix the xfail
  * [ ] Remove `test_forks.py`. (I'd like to go through it once more before that).
* [ ] Docs (i.e. updating the README and meta course).
* [x] Split out `naucse_render`

### Get a baseline from "almost master".
On a VM:
- pull my `sanitize-absolute-urls` branch (see #504)
```console
$ git checkout sanitize-absolute-urls
$ ARCA_BACKEND=arca.backend.DockerBackend FORKS_ENABLED=true ARCA_BACKEND_REGISTRY_PULL_ONLY=1 TRAVIS=1 TRAVIS_REPO_SLUG=pyvec/naucse.python.cz TRAVIS_BRANCH=master NAUCSE_CALENDAR_DTSTAMP=2018-11-22 pipenv run naucse freeze -v
$ mv naucse/_build/ _build_master
```
### Clean the caches
just to be sure:
```console
$ rm -rfv .arca/
$ docker stop $(docker ps -a -q) && docker container prune -f
```

### Freeze
* [x] Merge https://github.com/muzikovam/naucse.python.cz/pull/49
* pull this PR's fork
```console
$ git checkout fork-api
$ NAUCSE_MAIN_REPO_URL=https://github.com/pyvec/naucse.python.cz/ NAUCSE_CALENDAR_DTSTAMP=2018-11-22 NAUCSE_TRUSTED_REPOS='*' ARCA_BACKEND=arca.backend.DockerBackend ARCA_BACKEND_REGISTRY_PULL_ONLY=1 pipenv run naucse freeze -v 
```

### Compare
```console
$ YOUR_FAVORITE_RECURSIVE_DIFF_TOOL="diff -r -U3"
$ $YOUR_FAVORITE_RECURSIVE_DIFF_TOOL _build_master/ naucse/_build
```
For me, the unified diff has 4649 lines :(
But it's mostly trivial differences – in things that shouldn't be rendered in forks, but were.

### CODE REVIEW

**Comment by hroncok** [comment]
For starters, the build diff looks reasonable.

File: naucse/edit_info.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+    # Otherwise, fake it!
+    assumed_url = os.environ.get('NAUCSE_MAIN_REPO_URL')
+    if assumed_url:
+        return get_repo_info(
+            assumed_url,
+            os.environ.get('NAUCSE_MAIN_REPO_BRANCH', 'master'),
+        )
+
+    return LocalRepoInfo(path)
```
Comment (hroncok) [inline naucse/edit_info.py]: --bookmark--

File: naucse/edit_info.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        return cls()
+
+    def get_edit_url(self, path):
+        return None
+
+
+class GithubRepoInfo(RepoInfo):
+    """Repository information for github.com"""
+
+    page_name='GitHubu'
```
Comment (hroncok) [inline naucse/edit_info.py]: as in „na GitHubu“?

**Comment by messa** [comment]
> @messa [v0.tar.gz](https://github.com/pyvec/naucse.python.cz/files/2746418/v0.tar.gz)
> 
> This is the API that gets published on the website. Could you please at least try to work with that to see if it plays nicely with https://projekty.pyladies.cz/ ?

BTW. the data that the [pyladies-courseware](https://github.com/messa/pyladies-courseware) project ([projekty.pyladies.cz](https://projekty.pyladies.cz/)) uses now are here: https://github.com/messa/pyladies-courseware/tree/master/courses But it is going to be refactored too then 😄 

I would like to fetch course + session structure from this API:
- list of courses (but the courseware will display only some of the courses, based on whitelist in courseware configuration)
- list of sessions in the course
- list of materials (just links, not the actual content) of each session. I thinks it's beneficial to have it in the courseware alongside the homeworks - people don't have to switch tabs/pages, search them etc.

I would keep the list of homeworks, their actual content and relation of homeworks to the sessions (or perhaps to the lessons?) in the courseware. At least for now. Primarily to keep things agile because there will be some homework content and metadata changes before the spring courses.

So now about the content of v0.tar.gz (thanks @hroncok for the file):

- `course.derives`, `course.vars` (for example in `v0/2018/pyladies-praha-podzim-ntk.json`) - just to be sure - courseware (projekty.pyladies.cz) doesn't need to use these values, because the list of materials for each session is already complete and I don't need to merge anything, is that right?
 
- `course.lessons.*.pages.*.solutions` - I guess this is out of scope of the courseware (projekty.pyladies.cz), actually I guess I don't need to work with `courses.lessons` at all. But I'm just curious, what are these standalone solution pages for? I see the solution is already included in the original page.

 - `course.sessions.*.pages` - so far I see that these pages contain just links to the materials. Or could there be any "non-index" page (page.url != session.url) with its own unique useful content?

- in the courseware, maybe it would be good idea to link the courseware homeworks to a lesson id rather than session id, and use the session-lesson info from this API. What is the lesson id - I guess the `lesson_slug` could work as a lesson id (at least in the course scope)?

- the `$ref` attributes are only present in `v0/naucse.json` and `v0/years/*.json`, not in course JSON (like `v0/2018/pyladies-praha-podzim-ntk.json`), is that right? I think the schema supports this assumption.

- just question: the `$ref` key (for example in `v0/naucse.json`) is something standardised? Looks interesting. I've found that it's standardised for the JSON schema documents for example.

**Comment by encukou** [comment]
> BTW. the data that the [pyladies-courseware](https://github.com/messa/pyladies-courseware) project ([projekty.pyladies.cz](https://projekty.pyladies.cz/)) uses now are here: https://github.com/messa/pyladies-courseware/tree/master/courses But it is going to be refactored too then smile
> 
> I would like to fetch course + session structure from this API:
> 
> * list of courses (but the courseware will display only some of the courses, based on whitelist in courseware configuration)
> 
> * list of sessions in the course
> 
> * list of materials (just links, not the actual content) of each session. I thinks it's beneficial to have it in the courseware alongside the homeworks - people don't have to switch tabs/pages, search them etc.

That's exactly what the API is for :)

> I would keep the list of homeworks, their actual content and relation of homeworks to the sessions (or perhaps to the lessons?) in the courseware. At least for now. Primarily to keep things agile because there will be some homework content and metadata changes before the spring courses.

Sounds great, just don't have the homework PDF links in courses that use courseware.

> So now about the content of v0.tar.gz (thanks @hroncok for the file):
> 
> * `course.derives`, `course.vars` (for example in `v0/2018/pyladies-praha-podzim-ntk.json`) - just to be sure - courseware (projekty.pyladies.cz) doesn't need to use these values, because the list of materials for each session is already complete and I don't need to merge anything, is that right?

Right, neither is useful to you. `course.derives` is mostly for backwards compatibility; `course.vars` is only used for rendering the content.

> * `course.lessons.*.pages.*.solutions` - I guess this is out of scope of the courseware (projekty.pyladies.cz), actually I guess I don't need to work with `courses.lessons` at all. But I'm just curious, what are these standalone solution pages for? I see the solution is already included in the original page.

Right, you don't need to use these.
They're for graceful degradation. Solutions are hidden – replaced with a link to the standalone page.
If you have JavaScript enabled, the solutions are just uncovered, so you won't see the standalone page.

> * `course.sessions.*.pages` - so far I see that these pages contain just links to the materials. Or could there be any "non-index" page (page.url != session.url) with its own unique useful content?

When some groups of attendees need different content, the `index` is a directory linking to other pages, as in: https://naucse.python.cz/course/pyladies/beginners/install/
Courseware should be OK with always linking to the index.

> * in the courseware, maybe it would be good idea to link the courseware homeworks to a lesson id rather than session id, and use the session-lesson info from this API. What is the lesson id - I guess the `lesson_slug` could work as a lesson id (at least in the course scope)?

Yes, `lesson_slug` is unique across a course.

> * the `$ref` attributes are only present in `v0/naucse.json` and `v0/years/*.json`, not in course JSON (like `v0/2018/pyladies-praha-podzim-ntk.json`), is that right? I think the schema supports this assumption.

Yes. The `$ref` links are part of the schema, so this won't change in stable API versions. `v0` is not stable, but I'll coordinate any changes with you.

> * just question: the `$ref` key (for example in `v0/naucse.json`) is something standardised? Looks interesting. I've found that it's standardised for the JSON schema documents for example.

It's taken from JSON schema. I don't know of a better standard for such linking (at least one suitable for the API being in static pages).

File: naucse/arca_renderer.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import shutil
+import re
+
+from arca import Task
+
+import naucse_render
+
+
+NAUCSE_URL_RE = re.compile(
+    r'^https://github.com/[^/]+/naucse\.python\.cz(\.git)?$'
```
[the remark this answers is not in the document]
Comment (encukou) [reply naucse/arca_renderer.py]: This is a hack.

I'd rather release `naucse_render` and add the dependency to all the courses (which we still can do).
Does the `naucse_render` code look OK? If yes I'll go and do that.

File: naucse/converters.py [hunk 4]
```diff
@@ -0,0 +1,614 @@
...
+        are supported.
+
+        The base implementation returns `data` unchanged.
+        """
+        return data
+
+    def dump(self, value):
+        """Convert a Python value to JSON-compatible data.
+
+        The base implementation returns `data` unchanged.
```
[the remark this answers is not in the document]
Comment (encukou) [reply naucse/converters.py]:
```suggestion
        The base implementation returns `value` unchanged.
```

File: naucse/converters.py [hunk 5]
```diff
@@ -0,0 +1,614 @@
...
+    This class is designed to be subclassed. Subclasses may override:
+
+    `load`, `dump`, `get_schema` (see docstrings)
+
+    `load_arg_names`: Names of keyword arguments for the `load` method.
+        Collection converters (for list & dict) pass these through.
+
+    `get_schema_url`: A method that determines the schema URL for a given
+        instance.
+        If it returns None, values cannot dumped using the top-level
```
[the remark this answers is not in the document]
Comment (encukou) [reply naucse/converters.py]: GitHub's "apply suggestions" [Beta] didn't list you as co-author. Reported. Sorry!

**Comment by encukou** [comment]
> The converter stuff seems like an external library material

Sadly, I didn't find one I'd want to use here. I did study `attrs` quite a lot, and read the docs of Colander and several [attrs extensions](https://github.com/python-attrs/attrs/wiki/Extensions-to-attrs). I had mainly two issues:
* I really want a language-agnostic schema.
* I didn't find enough focus on schema versions. Granted, this PR doesn't really do that either, but unlike with an external library the support can be added.

I'm not ready to support `converters.py` as an external library for other use cases, but it would be nice to do one day. That's why it's separated from `models`.

File: naucse_render/course.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    elif isinstance(value, dict):
+        return {k: encode_dates(v) for k, v in value.items()}
+    elif isinstance(value, (list, tuple)):
+        return [encode_dates(v) for v in value]
+    elif isinstance(value, (str, int, bool, type(None))):
+        return value
+    raise TypeError(value)
+
+
+# XXX: Clarify the version
```
Comment (hroncok) [inline naucse_render/course.py]: what version? it says `API_VERSION = 1` above.

File: naucse_render/course.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+def get_course(course_slug: str, *, path='.', version):
+    """Get information about the course as a JSON-compatible dict"""
+
+    base_path = Path(path).resolve()
+
+    if version <= 0:
```
Comment (hroncok) [inline naucse_render/course.py]: What version is this? The API_VERSION above? Shouldn't it be used here?

```suggestion
    if version < API_VERSION:
```

File: naucse_render/course.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        raise ValueError(f'Version {version} is not supported')
+
+    if course_slug == 'lessons':
```
Comment (hroncok) [inline naucse_render/course.py]: I'd actually make this a function, and add a docstring. It's hard to grasp and I had to walk the code in my head to understand *what* it does. Knowing *why* would be even better.

File: naucse_render/course.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+    elif isinstance(value, (list, tuple)):
+        return [encode_dates(v) for v in value]
+    elif isinstance(value, (str, int, bool, type(None))):
+        return value
+    raise TypeError(value)
+
+
+# XXX: Clarify the version
+
+def get_course(course_slug: str, *, path='.', version):
```
Comment (hroncok) [inline naucse_render/course.py]: This function generally is very long and sometimes very "weird". Aka it says "rename x to y" etc., but it is not clear why.
Is this to convert content form before this PR to the new internal API spec?

File: naucse_render/course.py (hunk 8 above)
Comment (hroncok) [reply naucse_render/course.py]: (actually, see the more general comment near the header of this function)

File: naucse_render/course.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+            if page_path.exists():
+                session['pages'] = {}
+                for page_md_path in page_path.glob('*.md'):
+                    session['pages'][page_md_path.stem] = {
+                        'content': convert_markdown(page_md_path.read_text()),
+                    }
+
+    result = encode_dates(info)
+    return {
+        'api_version': [0, 0],
```
Comment (hroncok) [inline naucse_render/course.py]: what does this actually mean?

Comment (encukou) [reply naucse_render/course.py]: 0.0, just in a format that's easier to parse.
Do you think that's a bad idea?

File: naucse_render/course.py (hunk 9 above)
Comment (encukou) [reply naucse_render/course.py]: Yes.
I added some comments.

File: naucse_render/course.py (hunk 6 above)
Comment (encukou) [reply naucse_render/course.py]: Ah, right. That's a remnant of when I tried to add versioning as well.
The PR is big enough without that, so I'll leave it to the future.

File: naucse_render/course.py (hunk 10 above)
Comment (hroncok) [reply naucse_render/course.py]: the comment helped.

File: naucse/arca_renderer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (encukou) [reply naucse/arca_renderer.py]: Hack removed. Use `NAUCSE_TRUSTED_REPOS='https://github.com/encukou/naucse.python.cz.git#beginners-2018-exp'` until `naucse_render` is added to forks.
Should I update the forks now, or do you want to test more?

**Comment by hroncok** [comment]
The tests forks are erroring, as their shared fixture no longer works. I suggest we remove those and create a new set of tests that tests the following scenarios:

* [ ] A valid trvial json that could come from a fork is transformed correctly into the model
* [ ] A valid nontrvial json that could come from a fork is transformed correctly into the model
* [ ] An invalid json that could come from a fork is not transformed into the model and a proper exception happens
* [ ] A fork is made (could be adapted from the fixture), data is loaded, it loads the correct data.
* [ ] A fork is made, yet it has a yaml error, data is not loaded, an exception that makes sense happens instead. The consumer of the exception is aware that this happened due to a YAML error in XY file in ABC fork. This needs to be visible in the travis log  in case it happens in reality.
* [ ] A fork is made, yet the fork doesn't have naucse_renderer in Pipfile. The exception mentions that.
* [ ] A fork wants a newer version of our API that we don't yet have. We check that the error mentions that.
* [ ] A fork deletes all the data. Note sure what happens.
* [ ] A broken fork link.
* [ ] Test that NAUCSE_TRUSTED_REPOS is used and respected.
* [ ] Optionally unittest converters but that should be covered by some of high level tests above so this can remain as an open issue for "next" time.

**Comment by hroncok** [comment]
The Travis freeze failure says: ModuleNotFoundError: No module named 'naucse_render'

That goes from a fork i guess, but how can I see which one?

**Comment by encukou** [comment]
> The Travis freeze failure says: ModuleNotFoundError: No module named 'naucse_render'
> 
> That goes from a fork i guess, but how can I see which one?

[The traceback](https://travis-ci.org/pyvec/naucse.python.cz/builds/481528946?utm_source=github_status&utm_medium=notification) now has a wrapper error with info on the repo and method called.
I went and [fixed that issue](https://github.com/cvut/naucse.python.cz/commit/ca7d8ee56b6997b83bb53f8b54876f1ebe375e18), a typo in the version, without review.

> A fork wants a newer version of our API that we don't yet have. We check that the error mentions that.

Versions are ignored entirely for now; there's no error and no bad input to check.

**Comment by hroncok** [comment]
> typo in the version, without review.

actually my bad review :( but at least we see it works :)

File: test_naucse/test_arca_render.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+
+def test_removed_data(arca_model, content_repo, git_command):
+    """Remove all data; check failing on FileNotFoundError"""
+    yaml_path = content_repo / 'courses/normal-course/info.yml'
+    run(
+        [git_command, 'rm', '-r', 'courses', 'runs', 'lessons'],
+        cwd=content_repo,
+    )
+    run([git_command, 'commit', '-a', '-m', 'Break YAML'], cwd=content_repo)
```
Comment (hroncok) [inline test_naucse/test_arca_render.py]:
```suggestion
    run([git_command, 'commit', '-a', '-m', 'Remove all data'], cwd=content_repo)
```

**Comment by encukou** [comment]
> Optionally unittest converters but that should be covered by some of high level tests above so this can remain as an open issue for "next" time.

Yup. Filed as https://github.com/pyvec/naucse.python.cz/issues/509.

> * [ ] A fork is made, yet the fork doesn't have naucse_renderer in Pipfile. The exception mentions that.

For some reason, I can't seem to get the right error in this case :/
Hopefully that's the last issue for tests. Docs are still remaining (i.e. updating the README and meta course). Anything else you can think of?

Signing out for today.

**Comment by hroncok** [comment]
> For some reason, I can't seem to get the right error in this case :/

what error you get? the [travis log](https://travis-ci.org/pyvec/naucse.python.cz/builds/481528946?utm_source=github_status&utm_medium=notification) from above shows what happens, doesn't it?

> Hopefully that's the last issue for tests. Docs are still remaining (i.e. updating the README and meta course). Anything else you can think of?

Good work with the tests, thank you. Any point of keeping the skipped test_forks? Do you have plans with them?

I must admit that I wasn't able to review everything so thoroughly as I'd like, but that would take me  a couple weeks. For others: I've consulted this PR with @encukou in person as well.

One problem I have is that everything is soooo complicated at this point that debugging any problem has a bus factor of 3 (@encukou @mikicz and me) and after this, @mikicz might no longer be counted into this. We shall not travel with the same train from now on :grin: 

I am afraid that at this point we are too deep in the rabbit hole and this PR doesn't make it that worse. Hopefully it *allows us* to make it better. Thanks for that.

> Signing out for today.

Good night :zzz:

**Comment by encukou** [comment]
>> For some reason, I can't seem to get the right error in this case :/
> 
> what error you get? the travis log from above shows what happens, doesn't it?

No error. If I delete the Pipfile/Pipfile.lock, and put in a requirements.txt without naucse_render, everything seems to work. And it's all hidden somewhere in Arca :(
Will look into it, but I need to do DevConf preparations this weekend.

>> Hopefully that's the last issue for tests. Docs are still remaining (i.e. updating the README and meta course). Anything else you can think of?
> 
> Good work with the tests, thank you. Any point of keeping the skipped test_forks? Do you have plans with them?

I'd like to go through it once more before removing it. Possibly after this PR is merged.

> I must admit that I wasn't able to review everything so thoroughly as I'd like, but that would take me a couple weeks. For others: I've consulted this PR with @encukou in person as well.
> 
> One problem I have is that everything is soooo complicated at this point that debugging any problem has a bus factor of 3 (@encukou @mikicz and me) and after this, @mikicz might no longer be counted into this. We shall not travel with the same train from now on grin
> 
> I am afraid that at this point we are too deep in the rabbit hole and this PR doesn't make it that worse. Hopefully it allows us to make it better. Thanks for that.

Indeed, that's the goal here. Not so much to reduce the overall complexity, but to separate it into individual pieces, so someone can focus on one thing (in naucse or outside), and ignore the others.

**Comment by mikicz** [comment]
> > > For some reason, I can't seem to get the right error in this case :/
> > 
> > 
> > what error you get? the travis log from above shows what happens, doesn't it?
> 
> No error. If I delete the Pipfile/Pipfile.lock, and put in a requirements.txt without naucse_render, everything seems to work. And it's all hidden somewhere in Arca :(
> Will look into it, but I need to do DevConf preparations this weekend.

I definitely can't read the whole dff/conversation, but I noticed this bit in a notification and I think I could take a short look if there's a problem with something in Arca or with Arca? Am I understanding correctly, that when you use the same fork with and without Pipfile/Pipfile.lock it produces a different result?

File: naucse/views.py [hunk 12]
```diff
@@ -1,119 +1,134 @@
...
 import datetime
-import logging
-import os
-from urllib.parse import urljoin
 from pathlib import Path
+import functools
+import calendar
+import os
 
+from flask import Flask, render_template, jsonify, url_for, Response, abort, g
```
Comment (mikicz) [inline naucse/views.py]:
```suggestion
from flask import Flask, render_template, jsonify, url_for, Response, abort, g, redirect
```

The function is used in the runs route, but not imported.

Comment (hroncok) [reply naucse/views.py]: Nice catch. I think the redirect there is not very useful and we should remove that call instead. Teh website is frozen, this doesn't work on production.

**Comment by encukou** [comment]
> I definitely can't read the whole dff/conversation, but I noticed this bit in a notification and I think I could take a short look if there's a problem with something in Arca or with Arca? Am I understanding correctly, that when you use the same fork with and without Pipfile/Pipfile.lock it produces a different result?

Sorry for the delay.

I've added the failing test in https://github.com/pyvec/naucse.python.cz/pull/505/commits/1162e6504b2d4e827c36a80954756446cc3938a3
Hopefully it's clear enough. There are a few variations I've used, but I'm not sure how to make it fail.

**Comment by encukou** [comment]
The API is blocking two external projects (coursweare and feedbacker).
Let us merge, and put remaining work (tests, docs) in issues. The whole thing is a big integration test – and that passes.

**Comment by encukou** [comment]
It's [alive](https://naucse.python.cz/v0/naucse.json)! Thanks everyone for your help!

TODO items merged into https://github.com/pyvec/naucse.python.cz/issues/494

**Pull request merged**